packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -119,29 +124,137 @@ export function publish(options: InitOptions): void {
+    observer.unobserve(element);
+    observedRoots.delete(element, surfaceData);
+    surfaceDataRoots.delete(surfaceData, element);
+  }
+
+  // Handle an element by either observing its suitable children or setting up a mutation observer
+  function handleElementWithChildren(
+    observer: IntersectionObserver,
+    element: Element,
+    surfaceData: ALSurfaceData
+  ): void {
     const roots = getNonSurfaceWrapperRoots(element);
-    for (let i = 0; i < roots.length; ++i) {
-      const root = roots[i];
-      observer.observe(root);
-      observedRoots.set(root, surfaceData);
-      surfaceDataRoots.set(surfaceData, root);
+
+    if (roots.length > 0) {
+      // Normal case: we have viable children to observe
+      for (let i = 0; i < roots.length; ++i) {
+        trackElement(observer, roots[i], surfaceData);
+      }
+    } else if (enableDynamicChildTracking) {
+      // Special case: no viable children to observe, so set up a mutation observer
+      trackElement(observer, element, surfaceData, false);
+      // Set up mutation observer to watch for child additions
+      mutationObserver.observe(element, { childList: true });
     }
   }
+
+  const mutationObserver = new MutationObserver((mutations) => {
+    for (const mutation of mutations) {
+      if (mutation.type === 'childList') {
+        const element = mutation.target as Element;
+        const observer = getVirtualPropertyValue<IntersectionObserver>(element, VISIBILITY_OBSERVER_PROP);
+        const surfaceData = getVirtualPropertyValue<ALSurfaceData>(element, SURFACE_DATA_PROP);
+        if (!observer || !surfaceData) {
+          return;
+        }
+        // Handle removed nodes
+        if (mutation.removedNodes.length > 0) {
+          handleRemovedNodes(observer, element, surfaceData, mutation.removedNodes);
+        }
+
+        // Handle added nodes
+        if (mutation.addedNodes.length > 0) {
+          // Re-evaluate the entire element with its new children
+          const surfaceDataList = observedRoots.get(element);
+          if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+            // Untrack the parent element first
+            untrackElement(observer, element, surfaceData);
+            // Try to find suitable children with the updated DOM
+            const roots = getNonSurfaceWrapperRoots(element);
+            if (roots.length > 0) {
+              // Found suitable children, observe them
+              for (let i = 0; i < roots.length; ++i) {
+                trackElement(observer, roots[i], surfaceData);
+                break; // Only need to observe one viable child
+              }
+            } else {
+              // Still no suitable children, track the parent again
+              // Which will check for observable roots again, and attach a mutation observer if needed
+              trackElement(observer, element, surfaceData);
+            }
```

Comment:
call `handleElementWithChildren` but pass null for `mutationObserver`. This way, we ensure the same logic is applied all the time.

Reply:
I remember adding this `break;` in this block though which handleElementWithChildren doesn't include,  which was important I believe otherwise it would emit duplicate events.

Will have to retest it again,  it has been a bit but I believe that was needed